docs/src/design/decoder/constraints.md

```diff
@@ -0,0 +1,578 @@
+$$
+
+Using this variable, we can describe constraints against the $gc$ column as follows:
+
+Inside a *span* block, group count can either stay the same or decrease by one:
+
+> $$
+sp \cdot \Delta gc \cdot (\Delta gc - 1) = 0 \text{ | degree } = 3
+$$
+
+When group count is decremented inside a *span* block, either $h_0$ must be $0$ (we consumed all operations in a group) or we must be executing `PUSH` operation:
+
+> $$
+sp \cdot \Delta gc \cdot (1 - f_{push})\cdot h_0 = 0 \text{ | degree } = 5
+$$
+
+Notice that the above constraint does not preclude $f_{push} = 1$ and $h_0 = 0$ from being true at the same time. If this happens, op group decoding constraints (described [here](#op-group-decoding-constraints)) will force that the operation following the `PUSH` operation is a `NOOP`.
+
+When executing a `SPAN`, a `RESPAN`, or a `PUSH` operation, group count must be decremented by $1$:
+
+> $$
+(f_{span} + f_{respan} + f_{push}) \cdot (\Delta gc - 1) = 0 \text{ | degree } = 6
+$$
+
+If the next operation is either an `END` or a `RESPAN`, group count must remain the same:
+
+> $$
+\Delta gc \cdot (f_{end}' + f_{respan}') = 0 \text{ | degree } = 6
+$$
+
+When an `END` operation is executed, group count must be $0$:
+
+> $$
+f_{end} \cdot gc = 0 \text{ | degree } = 3
+$$
+
+### Op index constraints
+The `op_index` column (denoted as $ox$) tracks index of an operation within its operation group. It is used to ensure that the number of operations executed per group never exceeds $9$. The index is zero-based, and thus, the possible set of values for $ox$ is between $0$ and $8$ (both inclusive).
+
+To simplify the description of the constraints, we will define the following variables:
+
+$$
+ng = \Delta gc - f_{push} \\
+\Delta ox = ox' - ox
+$$
+
+The value of $ng$ is set to $1$ when we are about to start executing a new operation group (i.e., group count is decremented but we did not execute a `PUSH` operation). Using these variables, we can describe constraints against the $ox$ column as follows.
+
+When executing `SPAN` or `RESPAN` operations the next value of `op_index` must be set to $0$:
+
+> $$
+(f_{span} + f_{respan}) \cdot ox' = 0 \text{ | degree } = 6
+$$ 
+
+When starting a new operation group inside a *span* block, the next value of `op_index` must be set to $0$. Note that we multiply by $sp$ to exclude the cases when the group count is decremented because of `SPAN` or `RESPAN` operations:
+
+> $$
+sp \cdot ng \cdot ox' = 0 \text{ | degree } = 4
+$$
+
+When inside a *span* block but not starting a new operation group, `op_index` must be incremented by $1$. Note that we multiply by $sp'$ to exclude the cases when we are about to exit processing of an operation batch (i.e., the next operation is either `EDN` or `RESPAN`):
+
+> $$
+sp \cdot sp' \cdot (1 - ng) \cdot (\Delta ox - 1) = 0 \text{ | degree } = 5
+$$
+
+Values of `op_index` must be in the range $[0, 8]$.
+
+> $$
+\prod_{i=0}^{8}(ox - i) = 0 \text{ | degree } = 9
```

Comment:
For transition constraints degree 9 is fine because this refers only to the numerator part of the constraint (i.e., only the $c(x)$ part in $\frac{c(x)}{z(x)}$. Since for transition constraints the degree of $z(x)$ is about the same as the length of the execution trace, the overall degree of $\frac{c(x)}{z(x)}$ is still around 8.

Reply:
Ah, understood. Thanks!